forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.kt

```diff
@@ -0,0 +1,142 @@
+     */
+    suspend fun authenticate(context: Context,
+                             webAuthnKeySelector: WebAuthnKeySelector = WebAuthnKeySelector.DEFAULT): String {
+
+        var userHandle: ByteArray? = null
+        //username less when allowCredentials is empty
+        if (options.allowList.isNullOrEmpty()) {
+            //TODO What about username less with passkey?
+            val publicKeyCredentialSources = getPublicKeyCredentialSource(context)
+            //When there is only one stored credential, automatically trigger with the stored credential.
+            if (publicKeyCredentialSources.isNotEmpty()) {
+                //Launch a dialog and ask for which user for authentication
+                webAuthnKeySelector.select(publicKeyCredentialSources)?.let {
+                    options = options.cloneWith(listOf(it.toDescriptor()))
+                    userHandle = it.userHandle
+                }
+            }
+        }
+        return authenticate(context, userHandle)
+    }
+
+    /**
+     * Retrieve the [PublicKeyCredentialSource]
+     *
+     * @param context The Application Context
+     * @return The stored [PublicKeyCredentialSource]
+     */
+    internal open fun getPublicKeyCredentialSource(context: Context): List<PublicKeyCredentialSource> {
+        return WebAuthnDataRepository.builder()
+            .context(context).build()
+            .getPublicKeyCredentialSource(options.rpId)
+    }
+
+    private suspend fun authenticate(context: Context, userHandle: ByteArray?): String {
+        val publicKeyCredential = getPublicKeyCredential(context)
+        val response = publicKeyCredential.response as AuthenticatorAssertionResponse
+        val sb = StringBuilder()
+        sb.append(String(response.clientDataJSON))
+            .append("::")
+            .append(format(response.authenticatorData))
+            .append("::")
+            .append(format(response.signature))
+            .append("::")
+            .append(Base64.encodeToString(publicKeyCredential.rawId,
+                Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING))
+        if (userHandle != null) {
+            sb.append("::");
+            sb.append(Base64.encodeToString(userHandle, Base64.URL_SAFE or Base64.NO_WRAP))
+        } else {
+            if (response.userHandle != null) {
+                sb.append("::");
+                val decoded = Base64.decode(response.userHandle, Base64.DEFAULT)
+                sb.append(Base64.encodeToString(decoded, Base64.URL_SAFE or Base64.NO_WRAP))
+            }
+        }
+        return (sb.toString())
+    }
+
+    override suspend fun getPublicKeyCredential(context: Context): PublicKeyCredential {
+        val fido2ApiClient = Fido.getFido2ApiClient(context)
+        val task = fido2ApiClient.getSignPendingIntent(options)
+        task.await() //this may throw exception
+        return WebAuthnFragment.launch(pendingIntent = task.result!!)
```

Comment:
can the task.result be nullable ?

Reply:
No, it does not suppose to return null if successful, I have combined the await to avoid the !!